### PR TITLE
Making use of relative constraints 

### PR DESCRIPTION
**Pull request by daveshah1**
Not ready to be merged yet, still a bit rough around the edges and needs debugging to be taken out.

This PR replaces the current legalisation API with relative placement constraints being set in the packer and a generic legaliser for these constraints being used instead during placement. We also use "constraint distance" - the distance between current position of cells in a chain and satisfied relative constraints in that chain - as a SA metric

### CODE REVIEW

File: ice40/chains.h [hunk 1]
```diff
@@ -1,7 +1,7 @@
 /*
  *  nextpnr -- Next Generation Place and Route
  *
- *  Copyright (C) 2018  David Shah <david@symbioticeda.com>
+ *  Copyright (C) 2018  David Shah <clifford@symbioticeda.com>
```
Comment (eddiehung) [inline ice40/chains.h]: Make up your mind!

File: ice40/chains.cc [hunk 2]
```diff
@@ -0,0 +1,287 @@
+/*
+ *  nextpnr -- Next Generation Place and Route
+ *
+ *  Copyright (C) 2018  David Shah <clifford@symbioticeda.com>
```
Comment (eddiehung) [inline ice40/chains.cc]: Copy-pasta fail here too

File: common/placer1.cc [hunk 3]
```diff
@@ -225,7 +225,8 @@ class SAPlacer
             // Once cooled below legalise threshold, run legalisation and start requiring
             // legal moves only
             if (temp < legalise_temp && !require_legal) {
-                legalise_design(ctx);
+                legalise_relative_constraints(ctx);
+                // FIXME
```
Comment (eddiehung) [inline common/placer1.cc]: What does this refer to?

File: common/placer1.cc [hunk 4]
```diff
@@ -60,11 +60,11 @@ class SAPlacer
                 type_idx = bel_types.at(type);
             }
             if (int(fast_bels.size()) < type_idx + 1)
-                fast_bels.resize(type_idx + 1);
+            fast_bels.resize(type_idx + 1);
```
Comment (eddiehung) [inline common/placer1.cc]: clangformat necessary?

File: common/placer1.cc (hunk 3 above)
Comment (daveshah1) [reply common/placer1.cc]: Obsolete, from before I put the legalise step in there again

**Pull request merged**